### PR TITLE
Merge chisq + polchisq dispatchers into unified term-level family

### PR DESCRIPTION
**Pull request by rohandahale**
Closes task **2.11** in #212.

Merges the dual `chisq` + `polchisq` dispatchers into a single term-level family in `imager_backend.py`:

- **`compute_chisq_term`**, **`compute_chisqgrad_term`**, **`compute_chisqdata_term`** — dict-dispatched on `(dtype, ttype)`. 12 dtypes × valid ttypes, parity-tested bit-identically against the legacy paths.
- **`_pol_solve_block(which_solve, pol)`** helper encapsulates the `which_solve[0:4]` slicing pattern flagged as fragile in #227 (with a TODO for the Phase 6 `WhichSolve(stokes, spectral)` NamedTuple refactor).
- **Drops the `pol_next='I'` override** in `compute_data_tuples` flagged in #232 review; the validation moved into `compute_chisqdata_term`.
- **Legacy `chisq`/`chisqgrad`/`chisqdata` + `polchisq`/`polchisqgrad`/`polchisqdata`** become thin shims around the new term functions — back-compat preserved for `Obsdata.chisq`/`polchisq`, plotting, and calibration.
- **Cycle-break**: `embed_imarr` moved from `imager_backend.py` to `imager_utils.py` (next to `embed`) and vectorized over the `nsolve` axis. Bit-identical output to the previous per-row loop.

## JAX-readiness

`_CHISQ_DISPATCH` and `_CHISQDATA_DISPATCH` are the Phase 5 entry points: each entry can be swapped in place to register a JAX backend (e.g. `_CHISQ_DISPATCH['vis']['direct'] = (jax_chisq_vis, jax_chisqgrad_vis)`). `compute_chisq_term` is `jax.jit`-friendly under `static_argnames=('dtype', 'ttype')`.

## Design notes

- `compute_chisqgrad_term` requires explicit `pol_solve` for pol dtypes — no hidden default. The legacy shim still defaults to `POL_SOLVE_DEFAULT` and forwards it through, so external callers of `polchisqgrad(...)` without `pol_solve` are unchanged.
- Pol direct chisqdata leaves (`chisqdata_pvis`/`m`/`vvis`) now accept `**kwargs` so the unified dispatcher passes standard kwargs uniformly across all dtypes.

### CODE REVIEW

File: ehtim/imaging/imager_utils.py [hunk 1]
```diff
@@ -368,69 +222,19 @@ def regularizergrad(imvec, nprior, mask, flux, xdim, ydim, psize, stype, **kwarg
...
+    """Return (data, sigma, A) for a standard data term.
+
+    Thin shim around imager_backend.compute_chisqdata_term retained for
+    backward compatibility. New code should call compute_chisqdata_term
+    directly. Returns (False, False, False) (legacy silent fallback) if
+    `dtype` is not a standard data term.
+    """
+    from ehtim.imaging.imager_backend import compute_chisqdata_term
+    ttype = kwargs.pop('ttype', 'direct')
+    if dtype not in DATATERMS:
```
Comment (achael) [inline ehtim/imaging/imager_utils.py:231]: I think it is fine to raise an explicit exception here, I don't think this fallback is ever used.

File: ehtim/imaging/imager_utils.py [hunk 2]
```diff
@@ -42,179 +42,33 @@
...
-    if mask is None:
-        mask = []
-    chisqgrad = np.zeros(len(imvec))
+    Thin shim around imager_backend.compute_chisqgrad_term retained for
+    backward compatibility. New code should call compute_chisqgrad_term
+    directly. Returns zeros (legacy silent fallback) if `dtype` is not a
+    standard data term.
+    """
+    from ehtim.imaging.imager_backend import compute_chisqgrad_term
     if dtype not in DATATERMS:
```
Comment (achael) [inline ehtim/imaging/imager_utils.py:66]: maybe better to put in an explicit exception, I don't think this fallback is used.

**Comment by rohandahale** [comment]
@achael resolving merge conflicts now

File: ehtim/imaging/imager_utils.py [hunk 3]
```diff
@@ -42,179 +42,33 @@
...
-    if mask is None:
-        mask = []
-    chisq = 1
+    Thin shim around imager_backend.compute_chisq_term retained for backward
+    compatibility. New code should call compute_chisq_term directly. Returns
+    1 (legacy silent fallback) if `dtype` is not a standard data term.
+    """
+    # Imported here to avoid a module-load cycle with pol_imager_utils.
+    from ehtim.imaging.imager_backend import compute_chisq_term
     if dtype not in DATATERMS:
```
Comment (achael) [inline ehtim/imaging/imager_utils.py:52]: maybe better to put in an explicit exception

File: ehtim/imaging/imager_backend.py [hunk 4]
```diff
@@ -536,6 +508,282 @@ class ImagerInitState(NamedTuple):
     reffreq: float               # may be re-bound to init.rf when mf=True
 
 
+# `logamp` aliases `amp` (same data product). Pol dtypes have no `fast`.
+_CHISQDATA_DISPATCH = {
```
Comment (achael) [inline ehtim/imaging/imager_backend.py]: This looks great, but maybe it makes sense to reorder the imager_backend file so this is at the top of the follow, followed by classes, followed by functions?

**Comment by rohandahale** [comment]
@achael The conflicts are resolved now

File: ehtim/imaging/imager_backend.py [hunk 5]
```diff
@@ -536,6 +508,282 @@ class ImagerInitState(NamedTuple):
...
+    # Standard data terms consume a single Stokes letter ('I'/'Q'/'U'/'V'). When
+    # the imager is in a multi-Stokes mode (e.g. 'IP', 'IQUV'), the leaf needs
+    # pol='I' to unpack Stokes-I visibilities; pol modes without 'I' (e.g. 'P')
+    # are incompatible with standard data terms.
+    if not is_pol and pol in POLARIZATION_MODES:
+        if 'I' not in pol:
+            raise Exception(f"cannot use dterm {dtype} with pol={pol}")
+        pol = 'I'
+
+    # Standard direct + all pol leaves take mask positionally.
```
Comment (achael) [inline ehtim/imaging/imager_backend.py]: polarization chisqdata functions with nfft/fast don't actually use mask -- maybe we can remove the arguments in pol_imager_utils so this is more symmetric and the only case when a mask is passed is 'direct'?

File: ehtim/imaging/imager_backend.py [hunk 6]
```diff
@@ -536,6 +508,282 @@ class ImagerInitState(NamedTuple):
...
+    # Standard fast/nfft leaves omit it (they index uv coords from obs directly).
+    if is_pol or ttype == 'direct':
+        return helper(obs, prior, mask, pol=pol, **kwargs)
+    return helper(obs, prior, pol=pol, **kwargs)
+
+
+# _diag leaves index the gridded image directly per-baseline; their `fast`
```
Comment (achael) [inline ehtim/imaging/imager_backend.py]: again i'd move these global dispatch variables to the top of the file or a separate file

File: ehtim/imaging/imager_backend.py [hunk 7]
```diff
@@ -536,6 +508,282 @@ class ImagerInitState(NamedTuple):
...
+    if ttype not in entry:
+        raise Exception(f"ttype={ttype!r} not supported for dtype={dtype!r}")
+    chisq_fn, chisqgrad_fn = entry[ttype]
+    return chisq_fn, chisqgrad_fn, entry['is_pol']
+
+
+def compute_chisq_term(imcur, dtype, A, data, sigma, ttype='direct', mask=None):
+    """Single chi^2 dispatcher unifying chisq + polchisq.
+
+    Pol dtypes consume the full multi-row imcur ((4, npix) single-frequency
```
Comment (achael) [inline ehtim/imaging/imager_backend.py]: don't think the top part of this docstring is clarifying, i would just remove it. i think the description under imcur in the parameters does the job better of saying that the shape of imcur changes depending on the mode

File: ehtim/imaging/imager_backend.py [hunk 8]
```diff
@@ -536,6 +508,282 @@ class ImagerInitState(NamedTuple):
...
+
+    Returns
+    -------
+    float
+    """
+    chisq_fn, _, is_pol = _lookup_chisq_entry(dtype, ttype)
+
+    imvec = imcur if (is_pol or imcur.ndim == 1) else imcur[0]
+
+    # randomfloor=True breaks TV-regularizer gradient singularities.
```
Comment (achael) [inline ehtim/imaging/imager_backend.py]: I think the newer method of breaking up the TV singularities (epsilon_tv) should also be able to help with embedded cases without requiring the randomfloor. For now, can we define a global at the top e.g. EMBED_RANDOMFLOOR=True with a note to revisit, instead of setting this as True here?

File: ehtim/imaging/imager_backend.py [hunk 9]
```diff
@@ -536,6 +508,282 @@ class ImagerInitState(NamedTuple):
...
+    if (ttype != 'direct' and mask is not None
+            and len(mask) > 0 and np.any(np.invert(mask))):
+        if is_pol:
```
Comment (achael) [inline ehtim/imaging/imager_backend.py:928]: let's put  a TODO at the top to consolidate embed_imarr and embed, I can take a look at this later

File: ehtim/imaging/pol_imager_utils.py [hunk 10]
```diff
@@ -345,91 +345,33 @@ def vcv_grad(imarr, gradarr):
...
-    """
+    """Return chi^2 for a polarimetric data term.
 
-    chisq = 1
+    Thin shim around imager_backend.compute_chisq_term retained for backward
+    compatibility. New code should call compute_chisq_term directly.
+    Returns 1 (legacy silent fallback) if `dtype` is not a polarimetric term.
+    """
+    from ehtim.imaging.imager_backend import compute_chisq_term
     if dtype not in DATATERMS_POL:
```
Comment (achael) [inline ehtim/imaging/pol_imager_utils.py:354]: maybe better to put in explicit exception

File: ehtim/imaging/pol_imager_utils.py [hunk 11]
```diff
@@ -345,91 +345,33 @@ def vcv_grad(imarr, gradarr):
...
-    """return the chi^2 gradient for the appropriate dtype
+    Thin shim around imager_backend.compute_chisqgrad_term retained for
+    backward compatibility. New code should call compute_chisqgrad_term
+    directly. Returns zeros (legacy silent fallback) if `dtype` is not a
+    polarimetric term.
     """
-
-    chisqgrad = np.zeros(imarr.shape)
+    from ehtim.imaging.imager_backend import compute_chisqgrad_term
     if dtype not in DATATERMS_POL:
```
Comment (achael) [inline ehtim/imaging/pol_imager_utils.py:369]: explicit exception again

File: ehtim/imaging/pol_imager_utils.py [hunk 12]
```diff
@@ -525,36 +467,19 @@ def polregularizergrad(imarr, priorarr, mask, flux, pflux, vflux, xdim, ydim, ps
...
-            (data, sigma, A) = chisqdata_pvis_nfft(Obsdata, Prior, mask, **kwargs)
-        elif dtype == 'm':
-            (data, sigma, A) = chisqdata_m_nfft(Obsdata, Prior, mask, **kwargs)
-        elif dtype == 'vvis':
-            (data, sigma, A) = chisqdata_vvis_nfft(Obsdata, Prior, mask, **kwargs)
-
-    return (data, sigma, A)
+    from ehtim.imaging.imager_backend import compute_chisqdata_term
+    ttype = kwargs.pop('ttype', 'direct')
+    if dtype not in DATATERMS_POL:
```
Comment (achael) [inline ehtim/imaging/pol_imager_utils.py:476]: explicit exception

File: tests/test_imager_backend.py
Comment (achael) [inline tests/test_imager_backend.py:1]: this diff is too big for me to comment on on the github review page -- maybe there's another way to put these comments in I don't know about? If the file is getting unwieldy we could split it up into different types of tests

Comment (rohandahale) [reply tests/test_imager_backend.py:1]: Can you check if you can click on load diff and force load it?

File: ehtim/imaging/imager_utils.py (hunk 3 above)
Comment (rohandahale) [reply ehtim/imaging/imager_utils.py:52]: Done, now raises exception instead.

File: ehtim/imaging/imager_utils.py (hunk 2 above)
Comment (rohandahale) [reply ehtim/imaging/imager_utils.py:66]: Done!

File: ehtim/imaging/imager_utils.py (hunk 1 above)
Comment (rohandahale) [reply ehtim/imaging/imager_utils.py:231]: Done!

File: ehtim/imaging/pol_imager_utils.py (hunk 11 above)
Comment (rohandahale) [reply ehtim/imaging/pol_imager_utils.py:369]: Done!

File: ehtim/imaging/pol_imager_utils.py (hunk 12 above)
Comment (rohandahale) [reply ehtim/imaging/pol_imager_utils.py:476]: Done

File: ehtim/imaging/pol_imager_utils.py (hunk 10 above)
Comment (rohandahale) [reply ehtim/imaging/pol_imager_utils.py:354]: Done

File: ehtim/imaging/imager_backend.py (hunk 6 above)
Comment (rohandahale) [reply ehtim/imaging/imager_backend.py]: Done, `_CHISQDATA_DISPATCH`, `_DIAG_DTYPES`, `_CHISQ_DISPATCH`, and `ImagerInitState` all moved to the top of the file (order is constants --> dispatch tables --> class --> functions).

File: ehtim/imaging/imager_backend.py (hunk 7 above)
Comment (rohandahale) [reply ehtim/imaging/imager_backend.py]: Done, I removed the redundant intro, the `imcur` parameter description now has the shape  info.

File: ehtim/imaging/imager_backend.py (hunk 8 above)
Comment (rohandahale) [reply ehtim/imaging/imager_backend.py]: Done,`EMBED_RANDOMFLOOR = True` lives at the top with a comment noting `epsilon_tv` may make it obsolete, and both call sites use the constant.

File: ehtim/imaging/imager_backend.py (hunk 9 above)
Comment (rohandahale) [reply ehtim/imaging/imager_backend.py:928]: Added: `# TODO: consolidate `embed` (1D, this function) and `embed_imarr` (1D or 2D, below) into a single implementation -- their bodies overlap.`

File: ehtim/imaging/imager_backend.py (hunk 5 above)
Comment (rohandahale) [reply ehtim/imaging/imager_backend.py]: Done, `chisqdata_pvis_nfft`, `chisqdata_m_nfft`, `chisqdata_vvis_nfft` no longer take a `mask` arg, and the dispatcher in `compute_chisqdata_term` now only passes mask to `ttype == 'direct'`. Single check covers both standard fast/nfft (which never took mask) and pol nfft.

File: ehtim/imaging/imager_backend.py (hunk 4 above)
Comment (rohandahale) [reply ehtim/imaging/imager_backend.py]: Done, `_CHISQDATA_DISPATCH`, `_DIAG_DTYPES`, `_CHISQ_DISPATCH`, and `ImagerInitState` all moved to the top of the file (order is constants --> dispatch tables --> class --> functions).

File: tests/test_imager_backend.py
Comment (achael) [reply tests/test_imager_backend.py:1]: for some reason i thought that wasn't working before but it is now ...

**Comment by rohandahale** [comment]
@aviadlevis `embed_imarr` function is moved back to the imager due legacy dependencies but its computation is still in `imager_backend` (`compute_embed`). This will impact your jax backend file.

I investigated embed_imarr. The actual cycle is: `imager_backend` imports `pol_imager_utils` because `transform_imarr` needs polcv/mcv/vcv. Then `pol_imager_utils` needs `embed_imarr` (defined in `imager_backend`), but importing directly would be a hard cycle, so it goes through `ehtim.imager` as a middleman, `from ehtim.imager import embed_imarr`. This currently works only because ehtim/__init__.py's plotting/calibrating chain happens to pre-load pol_imager_utils before imager.py, so the partial import resolves. 

Fix: move polcv/mcv/vcv (and the _r and _grad variants) from `pol_imager_utils` to `imager_backend`. They're general image-basis transforms, not really pol-specific in who should own them. After that, `imager_backend` has zero upward deps, `pol_imager_utils` can import from it cleanly, and the back-door through `ehtim.imager` goes away. ~50 LOC move, no functional change.

**Comment by rohandahale** [comment]
@achael Should be ready to merge now. I resolved the merge conflicts.

**Review by achael — APPROVED** [review]
looks good, thanks for making the changes!
